### PR TITLE
introduce asynchronous plugin processing

### PR DESCRIPTION
**Pull request by steveej**
TODO
1. [x] use async for the plugin interface implementation
1. [x] port one plugin for testing
1. [x] ~~agree on the approach (the consumption and double-boxing of plugins for processing)~~. ~~allow mutable access to plugins for in-memory persistent plugin-state throughout the lifetime of the daemons~~. prefer a lock-frree read-only processing implementation and make the plugin responsible for handling their state via interior mutable fields. 
1. [x] remove all potential panics in non-test code (unwrap/expect)
1. [x] port all other plugins
1. [x] double-check and try to minimize the trait bounds
1. [x] clean up git history
1. [x] get #129 merged
1. [x] double-check for panics in non-test code
1. [x] ~discuss~ plugin initialization in graph-builder. https://github.com/openshift/cincinnati/pull/74 can be merged after this one.
1. [x] discuss `Box::leak` vs `lazy_static!`

### CODE REVIEW

**Comment by lucab** [comment]
@steveeJ If I'm reading this correctly, I'm in fact a bit concerned that we start consuming boxed-plugins instead of just referencing them. How do you envision glueing this together with the `Arc<Vec<BoxedPlugin>>` for configurable plugins from [here](https://github.com/openshift/cincinnati/pull/129/files#diff-f2a0e25396f0fcff0099ed163530e919R102)?

File: cincinnati/src/plugins/mod.rs [hunk 1]
```diff
@@ -295,45 +294,56 @@ where
...
+        )
     }
 }
 
 /// Processes all given Plugins sequentially.
 ///
 /// This function automatically converts between the different IO representations
 /// if necessary.
-pub fn process<T>(plugins: &[Box<T>], initial_io: InternalIO) -> Fallible<cincinnati::Graph>
+pub fn process<T>(plugins: T, initial_io: PluginIO) -> AsyncIO<InternalIO>
```
[the remark this answers is not in the document]
Comment (steveej) [reply cincinnati/src/plugins/mod.rs:326]: I wanted to ensure the parameters are passed along correctly too.

File: cincinnati/src/plugins/mod.rs [hunk 2]
```diff
@@ -295,45 +294,56 @@ where
...
-    let initial_io = PluginIO::InternalIO(initial_io);
-    let final_io = plugins
-        .iter()
-        .try_fold(initial_io, |last_io, next_plugin| next_plugin.run(last_io))?;
+    let future_result = futures::stream::iter_ok::<_, Error>(plugins)
+        .fold(
+            futures::future::Either::A(futures::future::ok::<_, Error>(initial_io)),
+            |future_io, next_plugin_mutex| {
+                next_plugin_mutex
+                    .lock()
```
[the remark this answers is not in the document]
Comment (steveej) [reply cincinnati/src/plugins/mod.rs]: > Is the mutex required for plugins that internally own and update mutable state?

Yes, that's the reason behind it.

> (Do we have any right now?)

They're not in the tree right now, but #138 introduces one.



> If so, do we maybe want to try pushing this one further down the line and use concurrent data-structures inside the plugins (i.e. atomic ints, concurrent maps) that can be internally mutated without locking the whole plugin?

By concurrent do you mean data structures which offer thread-safe interior mutability?

File: cincinnati/Cargo.toml [hunk 3]
```diff
@@ -14,9 +14,9 @@ protobuf = "2.0"
 url = "^1.7.2"
 log = "^0.4.3"
 futures = "0.1"
+futures-locks = "0.3.3"
 tokio = "0.1"
 toml = "^0.4.10"
-try_from = "0.3.2"
```
[the remark this answers is not in the document]
Comment (steveej) [reply cincinnati/Cargo.toml]: If it still helps I'll extract that into a different PR.

File: cincinnati/src/lib.rs [hunk 4]
```diff
@@ -47,8 +52,8 @@ pub const CONTENT_TYPE: &str = "application/json";
 const EXPECT_NODE_WEIGHT: &str = "all exisitng nodes to have a weight (release)";
 
 /// Graph type which stores `Release` as node-weights and `Empty` as edge-weights.
-#[derive(Debug, Default)]
-#[cfg_attr(test, derive(Clone))]
+#[derive(Debug, Default, Clone)]
```
[the remark this answers is not in the document]
Comment (steveej) [reply cincinnati/src/lib.rs]: Looks like this was a leftover! Removed, thanks :-)

File: cincinnati/src/plugins/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (lucab) [reply cincinnati/src/plugins/mod.rs]: > By concurrent do you mean data structures which offer thread-safe interior mutability?

Yes. That way each plugin is responsible of protecting only the portion of data-structures it needs to (possibly none, if those are atomic/lock-free) and we don't have to serialize plugins execution at a very coarse level.

[the remark this answers is not in the document]
Comment (steveej) [reply cincinnati/src/plugins/mod.rs]: > Yes. That way each plugin is responsible of protecting only the portion of data-structures it needs to (possibly none, if those are atomic/lock-free) and we don't have to serialize plugins execution at a very coarse level.

I acknowledge that there's some runtime overhead to locking the plugin. I'd be interested in the measurable influence of that. I don't have a very strong opinion and am willing to switch to `Plugin::run(&Self, ...)`, even though I do like the fact that the ease of mutating inside the plugins as it is now.

[the remark this answers is not in the document]
Comment (steveej) [reply cincinnati/src/plugins/mod.rs]: I switched to `&Self`, removed the mutex around the plugins and added an interior mutability example to the tests.

**Comment by steveej** [comment]
> discuss `Box::leak` vs `lazy_static!`

I [compared the alternatives](https://github.com/steveeJ/hello_rust/blob/940a5a4/static_refs/src/main.rs) and I think we would get away with the `Box::leak` approach in our code only instantiate the plugin vector once then use `Box::leak` to get and store a static reference.

File: cincinnati/src/plugins/macros.rs [hunk 5]
```diff
@@ -36,6 +36,18 @@ macro_rules! get_multiple_values {
     }
 }
 
+#[macro_export]
+macro_rules! new_plugin {
+    ($x:expr) => {
+        Box::new($x)
```
Comment (lucab) [inline cincinnati/src/plugins/macros.rs:42]: I think at this point we don't even need the macro anymore (but I'm fine either ways).

File: cincinnati/src/plugins/mod.rs [hunk 6]
```diff
@@ -281,45 +303,50 @@ where
...
-    }
-
-    let initial_io = PluginIO::InternalIO(initial_io);
-    let final_io = plugins
-        .iter()
-        .try_fold(initial_io, |last_io, next_plugin| next_plugin.run(last_io))?;
+    let future_result = futures::stream::iter_ok::<_, Error>(plugins)
+        .fold(initial_io, |io, next_plugin| next_plugin.run(io))
+        .into_future()
+        // .flatten()
```
Comment (lucab) [inline cincinnati/src/plugins/mod.rs]: Leftover.

File: policy-engine/src/main.rs [hunk 7]
```diff
@@ -67,7 +67,7 @@ fn main() -> Result<(), Error> {
         mandatory_params: settings.mandatory_client_parameters.clone(),
         upstream: settings.upstream.clone(),
         path_prefix: settings.path_prefix.clone(),
-        plugins: Arc::clone(&plugins),
+        plugins: Box::leak(Box::new(plugins)),
```
Comment (lucab) [inline policy-engine/src/main.rs:70]: `plugins` here is still an Arc. I think that isn't required anymore.

File: cincinnati/src/plugins/macros.rs (hunk 5 above)
Comment (steveej) [reply cincinnati/src/plugins/macros.rs:42]: That's true, it doesn't help if the type is stable. It was tremendously helpful during development. If there's only a slight chance that we'll change the type in the future, I'd prefer to keep it.

**Comment by steveej** [comment]
@lucab thanks for another review, comments addressed! PTAL

File: cincinnati/src/plugins/macros.rs (hunk 5 above)
Comment (lucab) [reply cincinnati/src/plugins/macros.rs:42]: Ack, fine for me.

**Pull request merged**